python_modules/dagster-test/dagster_test/toys/external_system_assets/__init__.py

```diff
@@ -0,0 +1,102 @@
+from dagster._core.definitions.assets import AssetsDefinition
+from dagster._core.definitions.data_version import DataProvenance, DataVersion
+from dagster._core.definitions.definitions_class import Definitions
+from dagster._core.definitions.events import AssetKey, Output
+from dagster._core.definitions.input import In
+from dagster._core.definitions.op_definition import OpDefinition
+from dagster._core.definitions.output import Out
+from dagster._core.types.dagster_type import Nothing
+from dagster._utils.backcompat import ExperimentalWarning
+
+from .external_system import ExternalSystem
+
+warnings.filterwarnings("ignore", category=ExperimentalWarning)
+
+EXTERNAL_SYSTEM = ExternalSystem()
+
+
+def external_asset(key: AssetKey):
+    def fn(context: OpExecutionContext, *args):
+        # Keep the temp storage directory used by the backend to the current instance.
+        EXTERNAL_SYSTEM.set_storage_directory(context.instance.storage_directory())
+
+        provenance = context.get_asset_provenance(key)
+        provenance_json = serialize_provenance(provenance) if provenance else None
+        data_version, is_memoized = EXTERNAL_SYSTEM.materialize(
+            key.to_user_string(),
+            provenance_json,
+        )
+        if is_memoized:
+            context.log.info(f"Used memoized value for {key.to_user_string()}")
+        yield Output(
+            None, data_version=DataVersion(data_version), metadata={"is_memoized": is_memoized}
+        )
+
+    deps = [AssetKey(k) for k in EXTERNAL_SYSTEM.get_dependencies(key.to_user_string())]
+    keys_by_input_name = {k.path[-1]: k for k in deps}
+    keys_by_output_name = {"result": key}
+    op_def = OpDefinition(
+        name=key.path[-1],
+        ins={k.to_user_string(): In(cast(type, Nothing)) for k in deps},
+        outs={"result": Out(Nothing)},
+        code_version=EXTERNAL_SYSTEM.get_code_version(key.to_user_string()),
+        compute_fn=fn,
+    )
+    return AssetsDefinition(
+        keys_by_input_name=keys_by_input_name,
+        keys_by_output_name=keys_by_output_name,
+        node_def=op_def,
+    )
+
+
+def serialize_provenance(provenance: DataProvenance) -> str:
+    return json.dumps(
+        {
+            "code_version": provenance.code_version,
+            "input_data_versions": {
+                key.to_user_string(): version.value
+                for key, version in provenance.input_data_versions.items()
+            },
+        }
+    )
+
+
+# ########################
+# ##### DEFINITIONS
+# ########################
+
+alpha = external_asset(AssetKey(["alpha"]))
+beta = external_asset(AssetKey(["beta"]))
+
+
+@observable_source_asset
+def delta(context) -> DataVersion:
+    EXTERNAL_SYSTEM.set_storage_directory(context.instance.storage_directory())
+    return DataVersion(EXTERNAL_SYSTEM.get_data_version("delta"))
+
+
+epsilon = external_asset(AssetKey(["epsilon"]))
+
+defs = Definitions(assets=[alpha, beta, delta, epsilon])
```

Comment:
could we include two jobs that materialize `alpha` and downstream, as well as `delta` and downstream?